### ISSUE TITLE
Share `coder_app` and ports with groups

### ISSUE THREAD

**Issue by stirby** [coder/coder#13361]
Currently, [`coder_app`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app)s can be [shared](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app#share) with either all authenticated users on a deployment or publicly. For some customers, these constraints are too loose. 

Often, IDEs or tools are configured for specific teams. For example, a data science team will want to share Jupyter notebooks with teammates, but not expose the sensitive data to all users on the deployment. 

To remedy this, we could add a new sharing mechanism to `coder_app` and sharable ports. There is no existing concept of groups in the terraform provider so this may warrant an RFC.

This is dependent on our ongoing "Coder as Code" work, and focus on it's implementation will likely be delayed to mid-Q3.

**Comment by stirby** [comment]
It would be nice if we could read users' groups in Terraform as a data source, since customers like to use generalized templates with team-specific options.

**Comment by matifali** [comment]
> It would be nice if we could access users' groups in Terraform (as a data source?) since customers like to use generalized templates with team-specific options.

It will be soon be possible to get a user's groups with `data.coder_workspace_owner.me.groups` having a list of all available groups would make the UX more useful. 

A good candidate for a feature request for https://github.com/coder/terraform-provider-coder

**Comment by stirby** [comment]
@matifali Interesting. I'll open a different issue there since this may impact shared ports as well. 

Do you envision that enabling this flow? Some conditional to check that the group matches?

```hcl
resource "coder_app" "code-server" {
  agent_id     = coder_agent.dev.id
  slug         = "code-server"
  ...
  count = contains(data.coder_workspace_owner.me.groups, "data-sci") ? 1 : 0
  share        = "authenticated"
}
```

If we could just pass the group name/id to the `share` attribute this would be extremely smooth. TBF this may be a more limiting abstraction.

```hcl
resource "coder_app" "code-server" {
  agent_id     = coder_agent.dev.id
  slug         = "code-server"
  ...
  share        = "group"
  group        = "data-sci" # or group id
}
```

**Comment by ammario** [comment]
I don't think that kind of string abuse is a good idea. In the `|` separator suggestion, a group named "public" or "owner" could not be selected. A user might try to target a group named "public"  and unknowingly make their shared URL public to the internet.

Also, we have tended to represent RBAC state in the DB as opposed to in Terraform. There are good reasons to do this consistently, e.g. when a Group is deleted or renamed we don't want Templates to fail.

Instead, I propose we add another share-level in between "owner" and "authenticated": "template". When a URL is shared to the "template", anyone that can use the template can also visit the URL.

**Comment by matifali** [comment]
`|` is not a separator here but just showing all mutually exclusive possible values for `share`. I agree that a change in group name can break the template. 
Using a new share level named `inherited` or `template` sounds a better approach.

**Comment by stirby** [comment]
> I don't think that kind of string abuse is a good idea. In the | separator suggestion, a group named "public" or "owner" could not be selected. A user might try to target a group named "public" and unknowingly make their shared URL public to the internet.

I was thinking about supporting group/user names/IDs directly but stopped because of this exact collision.

> Instead, I propose we add another share-level in between "owner" and "authenticated": "template". When a URL is shared to the "template", anyone that can use the template can also visit the URL.

I'm fine with this. 

Some customers still use ultra-general templates across multiple teams. However, I agree that we should signal templates-per-team.

**Comment by matifali** [comment]
If we limit the scope of this share to only ports shared through the dashboard. I think it's perfectly fine to share with an existing group or user by name. We can query and show all available users and groups to share to. 
This can also be limited to only share to groups and users who have access to the template, but I am not if that's what the users want.

**Comment by suse-coder** [comment]
I would also be interested in this feature

**Comment by matifali** [comment]
It would be ideal if we could do that interactively from the UI, where a user can share an app or a port with specific users, groups, or everyone.

**Comment by ericpaulsen** [comment]
Another customer request for this, surfaced during a trial with day-to-day workflow testing.

Use case: a developer wants to share a single forwarded port (e.g. a frontend dev server) with one colleague for review. Today the options are to share the port at `organization`/`authenticated` scope, which is broader than they want, or to share the whole workspace via workspace sharing, which grants SSH/terminal access they don't want to hand out just to show a UI.

What they're asking for is the flow described in https://github.com/coder/coder/issues/13361#issuecomment-2521369602: pick a port in the Open Ports UI and share it with specific users and/or groups, without touching workspace access.

_Posted by Coder Agents on behalf of @ericpaulsen._
